BootloaderCorePkg/Tools/SingleSign.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python
+## @ SingleSign.py
+# Single signing script
+#
+# Copyright (c) 2020, Intel Corporation. All rights reserved.<BR>
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+#
+##
+
+##
+# Import Modules
+#
+import os
+import sys
+import re
+import shutil
+import subprocess
+import struct
+import hashlib
+import string
+
+
+def get_openssl_path ():
+    if os.name == 'nt':
+        if 'OPENSSL_PATH' not in os.environ:
+            os.environ['OPENSSL_PATH'] = "C:\\Openssl\\"
+        if 'OPENSSL_CONF' not in os.environ:
+            openssl_cfg = "C:\\Openssl\\openssl.cfg"
+            if os.path.exists(openssl_cfg):
+                os.environ['OPENSSL_CONF'] = openssl_cfg
+    openssl = os.path.join(os.environ.get ('OPENSSL_PATH', ''), 'openssl')
+    return openssl
+
+def run_process (arg_list, print_cmd = False, capture_out = False):
+    sys.stdout.flush()
+    if print_cmd:
+        print (' '.join(arg_list))
+
+    exc    = None
+    result = 0
+    output = ''
+    try:
+        if capture_out:
+            output = subprocess.check_output(arg_list).decode()
+        else:
+            result = subprocess.call (arg_list)
+    except Exception as ex:
+        result = 1
+        exc    = ex
+
+    if result:
+        if not print_cmd:
+            print ('Error in running process:\n  %s' % ' '.join(arg_list))
+        if exc is None:
+            sys.exit(1)
+        else:
+            raise exc
+
+    return output
+
+#
+# Sign an file using openssl
+#
+# priv_key   [Input]        Key Id or Path to Private key
+# hash_type  [Input]        Signing hash
+# sign_scheme[Input]        Sign/padding scheme
+# in_file    [Input]        Input file to be signed
+# out_file   [Input/Output] Signed data file
+#
+
+def single_sign_file (priv_key, hash_type, sign_scheme, in_file, out_file):
+
+    _hash_type_string = {
+        "SHA2_256"    : 'sha256',
+        "SHA2_384"    : 'sha384',
+        "SHA2_512"    : 'sha512',
+    }
+
+    _hash_digest_Size = {
+        # Hash_string : Hash_Size
+        "SHA2_256"    : 32,
+        "SHA2_384"    : 48,
+        "SHA2_512"    : 64,
+        "SM3_256"     : 32,
+    }
+
+    _sign_scheme_string = {
+        "RSA_PKCS1"    : 'pkcs1',
+        "RSA_PSS"      : 'pss',
+    }
+
+
+    # Temporary files to store hash generated
+    hash_file_tmp = out_file+'.hash.tmp'
+    hash_file     = out_file+'.hash'
+
+    # Generate hash using openssl dgst in hex format
+    cmdargs = [get_openssl_path(), 'dgst', '-'+'%s' % _hash_type_string[hash_type], '-out', '%s' % hash_file_tmp, '%s' % in_file]
+    run_process (cmdargs)

```

Comment:
Does run_process throw user friendly error string if any of below conditions happens?
'priv_key' file is not present
'openssl.exe' is not present

Reply:
run_process() for openssl.exe not present - it would throw an exception. We would need to add additional checks for better handling.

For Priv_key not present - it would capture error.
FileNotFoundError: [Errno 2] No such file or directory: './BootloaderCorePkg/Tools/Keys/TestSigningPrivateKey1.pem'